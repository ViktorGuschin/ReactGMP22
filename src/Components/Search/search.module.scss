.searchText {
    color: #fff;
    text-transform: uppercase;
    font-weight: 300;
    font-size: 40px;
    line-height: 49px;
}

.searchInput {
    width: 713px;
    height: 57px;
    background: rgba(50, 50, 50, 0.8);
    mix-blend-mode: normal;
    opacity: 0.7;
    border: 0;
    border-radius: 4px;

    &::placeholder {
        font-size: 20px;
        line-height: 24px;
        color: #fff;
        mix-blend-mode: normal;
        opacity: 0.3;
        padding-left: 19px;
    }
}

.searchButton {
    width: 233px;
    height: 57px;
    margin-left: 14px;
    background-color: #f65261;
    border: 0;
    border-radius: 4px;
    cursor: pointer;
    font-weight: 500;
    font-size: 20px;
    line-height: 24px;
    text-transform: uppercase;
    color: #fff;
}

.searchInputWrapper {
    display: flex;
    align-items: center;
}
