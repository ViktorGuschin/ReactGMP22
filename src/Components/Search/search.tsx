import React, { useState } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';

import styles from './search.module.scss';
import useAppDispatch from '../../Hooks/useAppDispatch';
import { searchMoviesByTitle } from '../../Containers/MovieList/movieListSlice';

const Search: React.FunctionComponent = () => {
    // @ts-ignore
    const [searchText, setSearchText] = useState('');
    const navigate = useNavigate();
    // @ts-ignore
    const [searchParams] = useSearchParams();
    const dispatch = useAppDispatch();

    const handleOnChange = e => setSearchText(e.target.value);
    const handleSearch = () => {
        dispatch(searchMoviesByTitle(searchText));
        navigate(`/search/${searchText}?${searchParams}`);
    };

    return (
        <>
            <h2 className={styles.searchText}>find your movie</h2>
            <div className={styles.searchInputWrapper}>
                <input
                    className={styles.searchInput}
                    type="search"
                    value={searchText}
                    onChange={handleOnChange}
                    placeholder="What do you want to watch?"
                />
                <button className={styles.searchButton} type="button" onClick={handleSearch}>
                    search
                </button>
            </div>
        </>
    );
};

export default Search;
