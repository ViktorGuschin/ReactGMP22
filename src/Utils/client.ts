type ClientArgs = {
    body: any;
    headers: any;
    method: string;
};

const defaultConfig = {
    body: null,
    headers: null,
    method: 'GET',
};

const API_URL = 'http://localhost:4000';

export default async function client(endpoint, { body, ...customConfig }: ClientArgs = defaultConfig) {
    const headers = { 'Content-Type': 'application/json' };

    let config: ClientArgs = defaultConfig;
    if (body) {
        config.body = JSON.stringify(body);
    }
    config.method = body ? 'POST' : 'GET';
    config = { ...config, ...customConfig };
    config.headers = {
        ...headers,
        ...customConfig.headers,
    };

    let data;
    try {
        const response = await window.fetch(endpoint, config);
        data = await response.json();
        if (response.ok) {
            return data;
        }
        throw new Error(response.statusText);
    } catch (err) {
        return Promise.reject(err.message ? err.message : data);
    }
}

client.get = function (endpoint, customConfig = defaultConfig) {
    return client(`${API_URL}${endpoint}`, { ...customConfig, method: 'GET' });
};

client.post = function (endpoint, body, customConfig = defaultConfig) {
    return client(`${API_URL}${endpoint}`, { ...customConfig, body });
};

client.put = function (endpoint, body, customConfig = defaultConfig) {
    return client(`${API_URL}${endpoint}`, { ...customConfig, body, method: 'PUT' });
};

client.delete = function (endpoint, customConfig = defaultConfig) {
    return client(`${API_URL}${endpoint}`, { ...customConfig, method: 'DELETE' });
};
