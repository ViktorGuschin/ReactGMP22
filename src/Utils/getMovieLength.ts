const getMovieLength = (lengthInMinutes: number) => {
    const hours = Math.floor(lengthInMinutes / 60);
    const minutes = lengthInMinutes % 60;

    return `${hours}h ${minutes}min`;
};

export default getMovieLength;
