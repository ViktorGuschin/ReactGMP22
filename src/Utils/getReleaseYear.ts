const getReleaseYear = releaseDate => releaseDate.split('-')[0];

export default getReleaseYear;
