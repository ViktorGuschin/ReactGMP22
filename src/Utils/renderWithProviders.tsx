import React from 'react';
import { render } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
// As a basic setup, import your same slice reducers
import movieListReducer from '../Containers/MovieList/movieListSlice';
import deleteMovieModalReducer from '../Components/DeleteMovieModal/deleteMovieModalSlice';
import movieCardReducer from '../Components/MovieCard/movieCardSlice';
import addMovieModalReducer from '../Components/AddMovieModal/addMovieModalSlice';
import editMovieModalReducer from '../Components/EditMovieModal/editMovieModalSlice';

export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        store = configureStore({
            reducer: {
                movieList: movieListReducer,
                deleteMovieModal: deleteMovieModalReducer,
                addMovieModal: addMovieModalReducer,
                editMovieModal: editMovieModalReducer,
                movieCard: movieCardReducer,
            },
            preloadedState,
        }),
        ...renderOptions
    } = {},
) {
    function Wrapper({ children }) {
        return <Provider store={store}>{children}</Provider>;
    }

    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
