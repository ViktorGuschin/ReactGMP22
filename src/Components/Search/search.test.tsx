import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, screen } from '@testing-library/react';
import Search from './search';
import { renderWithProviders } from '../../Utils/renderWithProviders';

const navigateMock = jest.fn();
const useNavigateMock = () => jest.fn();

describe('Search Component', () => {
    beforeEach(() => {
        jest.mock('react-router-dom', () => ({
            ...jest.requireActual('react-router-dom'),
            useNavigate: useNavigateMock,
        }));
    });

    afterEach(() => {
        jest.unmock('react-router-dom');
    });

    it('should match snapshot', () => {
        const { asFragment } = renderWithProviders(
            <MemoryRouter>
                <Search />
            </MemoryRouter>,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should call with valid searchQuery by click button', () => {
        const value = 'test';
        renderWithProviders(
            <MemoryRouter>
                <Search />
            </MemoryRouter>,
        );

        const input = screen.getByPlaceholderText('What do you want to watch?');
        const button = screen.getByText('search');

        fireEvent.change(input, { target: { value } });
        fireEvent.click(button);

        expect(navigateMock).toHaveBeenCalledWith({ pathname: `/search/${value}`, search: '' });
    });
});
