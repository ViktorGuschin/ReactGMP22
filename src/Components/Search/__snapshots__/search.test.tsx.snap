// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component should match snapshot 1`] = `
<DocumentFragment>
  <h2
    class="searchText"
  >
    find your movie
  </h2>
  <div
    class="searchInputWrapper"
  >
    <input
      class="searchInput"
      placeholder="What do you want to watch?"
      type="search"
      value=""
    />
    <button
      class="searchButton"
      type="button"
    >
      search
    </button>
  </div>
</DocumentFragment>
`;
